Disable host process containers when disable unsafe operations is enabled

katiewasnothere opened the pull request:
Host process containers are not safe for multi-tenant scenarios since they run directly on the host and use the host's network. Don't allow host process containers to be used when `DisableUnsafeOperations` is set.

anmaxvl commented on pkg/annotations/annotations.go:
```diff
@@ -322,5 +322,6 @@ var AnnotationExpansions = map[string][]string{
 		WCOWDisableGMSA,
 		DisableWritableFileShares,
 		VSMBNoDirectMap,
+		HostProcessContainer,
```
shouldn't this be set to `false`, when disable unsafe annotations is `true`?

helsaawy replied:
oh yeah, i missed that
either an annotation for `DisableHostProcessContainer` needs to be created, or we need an `AnnotationDenyList` thingy

katiewasnothere replied:
@helsaawy I added a `DisableHostProcessContainer` annotation and updated the PR

katiewasnothere commented:
@helsaawy and @anmaxvl I reworked this PR to instead change how we handle annotation expanding. Instead of the annotations sharing the same value as the parent annotation, I added the ability to set a target value for the child annotations. Do y'all have any thoughts on this approach? 

Note: optionally in the future we could even expand this to use parent annotations that are not true or false. For example, I could envision an annotation like "memorybacking" and when that annotation's value is "physical" then it could expand with child annotations allowovercommit to false and vpmemdevicecount to 0, like we have for the annotation "fullyphysicallybacked"

helsaawy commented on pkg/annotations/annotations.go:
```diff
@@ -314,13 +314,13 @@ const (
-var AnnotationExpansions = map[string][]string{
+// other annotations and their values. It is an error for the expansions to already
+// exist and have a value that differs from the value set here.
+var AnnotationExpansions = map[string]map[string]string{
```
since we are already changing the type, which is breaking, can we make it:

```suggestion
func AnnotationExpansions() map[string]map[string]string {
```

that way, no-one can (inadvertently) modify the map

katiewasnothere replied:
Could we just make it a const instead?

helsaawy commented on internal/oci/annotations.go:
```diff
@@ -36,10 +36,15 @@ func ProcessAnnotations(ctx context.Context, s *specs.Spec) (err error) {
 		// check if annotation is present
-		if val, ok := s.Annotations[key]; ok {
+		if expVal, ok := s.Annotations[key]; ok {
+			// If the annotation is a bool and is set to false, we will treat that as
+			// equivalent to disabling the expansion.
+			if strings.ToLower(expVal) == "false" {
+				continue
+			}
```
this doesn't match the doc comment for `AnnotationExpansions`: 
```
It is an error for the expansions to already exist and have a value that differs from the value set here.
```
Also, we are assuming here that `AnnotationExpansions` has truthy semantics, when the comment and logic just assumes that the expansions have the same value as the original annotation

katiewasnothere replied:
I think I had changed that comment and what I meant by "a value that differs from the value set here" is that the child annotation's value set in the expansion definition cannot conflict with whatever is put on the spec. 

What do you mean by "truthy" semantics?

helsaawy commented on internal/oci/annotations_test.go:
```diff
@@ -55,24 +55,41 @@ func TestProccessAnnotations_Expansion(t *testing.T) {
+			if err != nil {
+				subtest.Fatalf("could not update spec from options: %v", err)
+			}
+			for k := range annotations.AnnotationExpansions[annotations.DisableUnsafeOperations] {
+				expected := annotations.AnnotationExpansions[annotations.DisableUnsafeOperations][k]
```
```suggestion
			for k, expected := range annotations.AnnotationExpansions[annotations.DisableUnsafeOperations] {
```

helsaawy reviewed (changes requested):
i think this implementation breaks the logic where we assume that parent annotation's must match the value of their expansions, and don't try to parse values as booleans.
i would rather we either:
 1. double down on the bool approach, and have `AnnotationExpansions` be `map[string]map[string]bool`; or
 2. add an `DisableHostProcessContainer` that is the antithesis of the `HostProcessContainer` annotation, with a separate check after annotation expansion that only one of the two can be defined

helsaawy reviewed (approved):
lgtm

anmaxvl commented on internal/oci/annotations.go:
```diff
@@ -55,6 +56,19 @@ func ProcessAnnotations(ctx context.Context, s *specs.Spec) (err error) {
 	}
 
+	// validate host process containers annotations are not conflicting
+	disableHPC := ParseAnnotationsBool(ctx, s.Annotations, annotations.DisableHostProcessContainer, false)
```
This is probably the only annotation that needs to be handled like this for now. In the future we may want to consider defining a "conflict annotations map" or similar, e.g.
```
conflictingAnnotations := map[string]string{
    annotations.DisableHostProcessContainer: annotations.HostProcessContainer,
}

for k, v := range conflictingAnnotations {
    v1 := ParseAnnotationsBool(ctx, s.Annotations, k, false)
    v2 := ParseAnnotationsBool(ctx, s.Annotations, v, false)
    if v1 && v2 {
        // handle conflict
    }
}
```

Maybe it's worth just doing it now too, rather than special casing it to HPC? thoughts?

anmaxvl reviewed (approved):
lgtm. my comment is non blocking
